specs/derivation.md

```diff
@@ -27,181 +43,606 @@ TODO
+containing when we start a channel, or even as we send the first frames in the channel.
+
+All of this is illustrated in the following diagram.
+
+> **TODO** improve diagram
+> - I'm a fan of the 4 lines "Transactions" to "L2 Blocks"
+>     - albeit it would good to show that channels & frames can occur out of order
+>     - but maybe that makes the diagram too hard to read and we can just include a comment afterwards saying that in
+>       general, reordering is possible (maybe show a second diagram showcasing a simple reordering?
+> - I think L1 blocks should be a new line above "Transactions" — also show deposits (w/ a number) for each block
+> - We shouldn't use the L1 attributes tx as a separate line, it makes it look like a additional layer of "data
+>   derivation", which it is not. Instead I would tag each L2 block with its epoch number.
+> - Include numbered deposits under L2 blocks
+> - Let's use a sequencing window of size 2 to keep the diagram small
+> - Include a textual explanation of the diagram below it
 
 ![](./assets/batch-deriv-chain.svg)
 
-### Frame format:
+### Batcher Transaction Format
 
 ```text
-frame = channel_id | frame_number | frame_data_length | frame_data | is_last
+transaction_data = version_byte ++ rollup_payload
+
+# version_byte == 0:
+rollup_payload = frame ...   #  one or more frames, concatenated
 
-channel_id        = bytes32 | uvarint   # identifies the channel to add frame_data to, as well as the time (UNIX-epoch in seconds) the channel was created.
-frame_number      = uvarint  # identifies the index of frame_data in the channel
-frame_data_length = uvarint  # frame_data_length == len(frame_data)
-frame_data        = bytes    # data to add to the channel, after frame max(frame_number-1, 0)
-is_last           = bool     # 1 byte, channel is closed if 1, stays open if 0, invalid otherwise
+# version_byte != 0: invalid, reserved for future use
 ```
 
-The channel identity uniqueness is limited both by the random `bytes32` in the ID, as well as the timestamp.
-The timestamp is limited by L1 timestamps, to force-close a channel and avoid revival of the same channel ID after the `CHANNEL_TIMEOUT`.
-This ensures at most a span of `CHANNEL_TIMEOUT` seconds of L1 data is required to reconstruct any channel-bank state in time.
+(where `++` denotes concatenation)
 
-### L1 transaction data format:
+The `rollup_payload` may be right-padded with 0s, which will be ignored. It's allowed for them to be
+interpreted as frames for channel 0, which must always be ignored.
+
+> **TODO** specify batcher authentication (i.e. where do we store / make available the public keys of authorize batcher
+> signers)
+
+### Frame Format
+
+A [channel frame][g-channel-frame] is encoded as:
 
 ```text
-transaction_data = version_byte | rollup_payload
+frame = channel_id ++ frame_number ++ frame_data_length ++ frame_data ++ is_last
+
+channel_id        = random ++ timestamp
+random            = bytes32
```

Comment:
Why `random`? Why not use an incrementing channel ID?

Reply:
Good question. I think we want to avoid relying on the number in the implem so that we're free to swap out the random value by a hash commitment in the future. @protolambda ?

Reply:
A nonce-like number would be stateful and thus much harder to recover (with current design there's no need to read L1 to make a valid batch-submission)